Content.Server/Nyanotrasen/StationEvents/Events/GlimmerShadeSpawnRule.cs

```diff
@@ -0,0 +1,57 @@
+using System.Linq;
+using Robust.Shared.Random;
+using Content.Server.GameTicking.Rules.Components;
+using Content.Shared.Atmos.Miasma;
+using Content.Server.StationEvents.Components;
+using Content.Shared.Psionics.Glimmer;
+
+namespace Content.Server.StationEvents.Events;
+
+internal sealed class GlimmerShadeSpawnRule : StationEventSystem<GlimmerShadeRuleComponent>
+{
+    [Dependency] private readonly IRobustRandom _robustRandom = default!;
+    [Dependency] private readonly GlimmerSystem _glimmerSystem = default!;
+
+    private static readonly string ShadePrototype = "MobShade";
+
+    protected override void Started(EntityUid uid, GlimmerShadeRuleComponent component, GameRuleComponent gameRule, GameRuleStartedEvent args)
+    {
+        base.Started(uid, component, gameRule, args);
+
+        var rottingPlaces = EntityManager.EntityQuery<RottingComponent, TransformComponent>().ToList();
+        var normalSpawnLocations = EntityManager.EntityQuery<VentCritterSpawnLocationComponent, TransformComponent>().ToList();
+        var hiddenSpawnLocations = EntityManager.EntityQuery<MidRoundAntagSpawnLocationComponent, TransformComponent>().ToList();
+
+        float chance = 0.15f * (float) _glimmerSystem.GetGlimmerTier();
+        int guaranteedSpawns = _robustRandom.Next(1, (int) _glimmerSystem.GetGlimmerTier() + 1);
+
+        // Spawn on top of rotting stuff first.
+        foreach(var (rot, xform) in rottingPlaces)
+        {
+            if (_robustRandom.Prob(chance))
+            {
+                EntityManager.SpawnEntity(ShadePrototype, xform.Coordinates);
+            }
+        }
+
+        int i = 0;
+        while (i < guaranteedSpawns)
+        {
+            if (normalSpawnLocations.Count != 0)
+            {
+                EntityManager.SpawnEntity(ShadePrototype, _robustRandom.Pick(normalSpawnLocations).Item2.Coordinates);
+                i++;
+                continue;
+            }
+
+            if (hiddenSpawnLocations.Count != 0)
+            {
+                EntityManager.SpawnEntity(ShadePrototype, _robustRandom.Pick(hiddenSpawnLocations).Item2.Coordinates);
+                i++;
+                continue;
+            }
+
+            return;
+        }
```

Comment:
If you do the HashSet union of TransformComponents (or maybe EntityCoordinates actually would be best since that's the only thing they're getting picked up for), you can rewrite this block to be a less computationally intensive and simpler.

You only need to check `Count` every iteration if you had plans to pop spawn locations off as they were being used. In that case, it might be better to shuffle a List, then iterate over the first N items as candidates, depending on the size of your List, or use the C# equivalent of a linked list with random popping.

```suggestion
        if (spawnLocations.Count == 0)
            return;

        while (guaranteedSpawns > 0)
        {
            --guaranteedSpawns;

            Spawn(ShadePrototype, _robustRandom.Pick(spawnLocations));
        }
```